src/parquet/file_reader.cc

```diff
@@ -0,0 +1,764 @@
+        }
+        int32_t len;
+        std::memcpy(&len, buffer.data(), 4);
+        if (len < 0) {
+            throw parquet_exception::corrupted_file("Negative rle encoding length");
+        }
+        if (static_cast<size_t>(len) > buffer.size()) {
+            throw parquet_exception::corrupted_file("Unexpected end of page");
+        }
+        _decoder = RleDecoder{buffer.data() + 4, len, _bit_width};
+        return 4 + len;
+    } else if (encoding == format::Encoding::BIT_PACKED) {
+        int64_t bit_len = static_cast<int64_t>(num_values) * _bit_width;
+        int64_t byte_len = (bit_len + 7) >> 3;
+        if (byte_len > std::numeric_limits<int>::max()) {
+            throw parquet_exception::corrupted_file("BIT_PACKED length exceeds int");
+        }
+        if (static_cast<size_t>(byte_len) > buffer.size()) {
+            throw parquet_exception::corrupted_file("Unexpected end of page");
+        }
+        _decoder = BitReader{buffer.data(), static_cast<int>(byte_len)};
+        return byte_len;
+    } else {
+        throw parquet_exception::nyi("Unknown level encoding");
+    }
+}
+
+void level_decoder::reset_v2(std::basic_string_view<uint8_t> encoded_levels, int num_values) {
+    assert(num_values >= 0);
+    _num_values = num_values;
+    _values_read = 0;
+    _decoder = RleDecoder{encoded_levels.data(), static_cast<int>(encoded_levels.size()), _bit_width};
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// plain_decoders
+
+template <typename T>
+void plain_decoder_trivial<T>::reset(std::basic_string_view<uint8_t> data) {
+    _buffer = std::move(data);
+}
+
+void plain_decoder_boolean::reset(std::basic_string_view<uint8_t> data) {
+    _decoder.Reset(const_cast<const uint8_t*>(data.data()), data.size());
+}
+
+void plain_decoder_byte_array::reset(std::basic_string_view<uint8_t> data) {
+    _buffer = seastar::temporary_buffer<uint8_t>(data.size());
+    std::memcpy(_buffer.get_write(), data.data(), data.size());
+}
+
+void plain_decoder_fixed_len_byte_array::reset(std::basic_string_view<uint8_t> data) {
+    _buffer = seastar::temporary_buffer<uint8_t>(data.size());
+    std::memcpy(_buffer.get_write(), data.data(), data.size());
+}
+
+template <typename T>
+size_t plain_decoder_trivial<T>::read_batch(size_t n, T out[]) {
+    size_t n_to_read = std::min(_buffer.size() / sizeof(T), n);
+    size_t bytes_to_read = sizeof(T) * n_to_read;
+    if (bytes_to_read > 0) {
+        std::memcpy(out, _buffer.data(), bytes_to_read);
+    }
+    _buffer.remove_prefix(bytes_to_read);
+    return n_to_read;
+}
+
+size_t plain_decoder_boolean::read_batch(size_t n, uint8_t out[]) {
+    return _decoder.GetBatch(1, out, n);
+}
+
+size_t plain_decoder_byte_array::read_batch(size_t n, seastar::temporary_buffer<uint8_t> out[]) {
+    for (size_t i = 0; i < n; ++i) {
+        if (_buffer.size() == 0) {
+            return i;
+        }
+        if (4 > _buffer.size()) {
```

Comment:
that's a Yoda condition - let's use human-readable `_buffer.size() < 4` instead

Reply:
Fixed in ac9cdae.